stlab/concurrency/router.hpp

```diff
@@ -0,0 +1,137 @@
+/*
+    Copyright 2015 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#ifndef STLAB_CONCURRENCY_ROUTER_HPP
+#define STLAB_CONCURRENCY_ROUTER_HPP
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <stlab/concurrency/channel.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+template <class T, class K>
+class router {
+    using channel_pair = std::pair<stlab::sender<T>, stlab::receiver<T>>;
+    using route_pair = std::pair<K, channel_pair>;
+    using routes = std::vector<route_pair>;
+
+public:
+    template <class E, class F>
+    router(E executor, F router_func) : _self{std::make_shared<model<E, F>>(std::move(executor), std::move(router_func))} {}
+
+    void set_ready() { _self->set_ready(); }
+    stlab::receiver<T> get_route(K key) { return _self->get_route(std::move(key)); }
+    stlab::receiver<T> get_default_route() { return _self->get_default_route(); }
+    void operator()(T t) { _self->route(std::move(t)); }
+
+private:
+    struct concept_t {
+        virtual stlab::receiver<T> get_route(K key) = 0;
+        virtual stlab::receiver<T> get_default_route() = 0;
+        virtual void set_ready() = 0;
+        virtual void route(T t) = 0;
+        virtual ~concept_t() = default;
+    };
+
+    template <class E, class F>
+    struct model : concept_t {
+        model(E executor, F router_func)
+        : _executor{std::move(executor)}
+        , _router_func{std::move(router_func)}
+        , _sender_receiver{stlab::channel<T>(_executor)}
+        {}
+
+        void set_ready() override {
+            std::sort(std::begin(_routes), std::end(_routes), [](const route_pair& a, const route_pair& b){
+                return a.first < b.first;
+            });
+            for (auto& pair : _routes)
+                pair.second.second.set_ready();
+            _sender_receiver.second.set_ready();
+        }
+
+        stlab::receiver<T> get_default_route() override { return _sender_receiver.second; }
+
+        stlab::receiver<T> get_route(K key) override {
+            auto find_it = std::find_if(begin(_routes), std::end(_routes), [&](const route_pair& pair){
+                return pair.first == key;
+            });
+            if (find_it != std::end(_routes)) return find_it->second.second;
+            _routes.push_back(std::make_pair(std::move(key), stlab::channel<T>(_executor)));
+            return _routes.back().second.second;
+        }
+
```

Comment:
I would remove this for the start of this router class. It makes the code more complicated for a use case, that I don't see really needed to be supported.